### PR TITLE
[IMP] stock: restricted access to scrap orders.

### PR DESCRIPTION
**Pull request by hdh-odoo**
**Description of the issue/feature this PR addresses**: 
- Task: https://www.odoo.com/web#id=1930055&action=333&active_id=131&model=project.task&view_type=form&menu_id=4720
- Pad: https://pad.odoo.com/p/r.ecf7eea9d9ccdf0d1c3e288e9757ef9c

**Current behavior before PR:** 
-Users can see scrap orders of all companies.

**Desired behavior after PR is merged:**
-Users cannot see scrap orders of companies other than their own.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/stock/models/stock_scrap.py [hunk 1]
```diff
@@ -25,6 +25,7 @@ def _get_default_location_id(self):
         'Reference',  default=lambda self: _('New'),
         copy=False, readonly=True, required=True,
         states={'done': [('readonly', True)]})
+    company_id = fields.Many2one('res.company', string='Company', default=lambda self: self.env.user.company_id)
```
Comment (mreficent) [inline addons/stock/models/stock_scrap.py]: maybe should be added `related='picking_id.company_id', readonly=True` 🤔

Comment (pedrobaeza) [reply addons/stock/models/stock_scrap.py]: In v12+, related are readonly by default.

File: addons/stock/models/stock_move_line.py [hunk 2]
```diff
@@ -79,7 +79,7 @@ def _check_lot_product(self):
     def _check_same_company_id(self):
         for line in self:
             if line.move_id and line.move_id.company_id != line.company_id:
-                raise ValidationError(_('The move line a a different company than its move'))
+                raise ValidationError(_('The move line has a different company than its move'))
```
Comment (Whenrow) [inline addons/stock/models/stock_move_line.py]: Could you move this diff in another commit ("fix typo ...")

File: addons/stock/models/stock_scrap.py (hunk 1 above)
Comment (Whenrow) [inline addons/stock/models/stock_scrap.py]: I would set this new field as required

File: addons/stock/models/stock_move_line.py (hunk 2 above)
Comment (hdh-odoo) [reply addons/stock/models/stock_move_line.py]: It's done. :)

File: addons/stock/models/stock_scrap.py (hunk 1 above)
Comment (hdh-odoo) [reply addons/stock/models/stock_scrap.py]: Done. Thanks.

File: addons/stock/models/res_company.py [hunk 3]
```diff
@@ -133,9 +148,21 @@ def _create_per_company_locations(self):
...
+
     @api.model
     def create(self, vals):
         company = super(Company, self).create(vals)
         company.sudo()._create_per_company_locations()
+        company.sudo()._create_scrap_sequence()
+        company.sudo()._create_transit_location()
+        company.sudo()._create_inventory_loss_location()
+        company.sudo()._create_production_location()
+        company.sudo()._create_scrap_location()
```
Comment (sle-odoo) [inline addons/stock/models/res_company.py]: nope

File: addons/stock/security/stock_security.xml [hunk 4]
```diff
@@ -158,5 +158,12 @@
         <field name="domain_force">['|',('company_id','=',False),('company_id','child_of',[user.company_id.id])]</field>
     </record>
 
+    <record model="ir.rule" id="stock_scrap_company_rule">
+        <field name="name">stock_scrap_company multi-company</field>
+        <field name="model_id" ref="model_stock_scrap"/>
+        <field name="global" eval="True"/>
+        <field name="domain_force">['|',('company_id','=',False),('company_id','child_of',[user.company_id.id])]</field>
```
Comment (sle-odoo) [inline addons/stock/security/stock_security.xml]: if company_id is requirent, the first part of the rule is useless

Comment (hdh-odoo) [reply addons/stock/security/stock_security.xml]: yes. removed that. Thanks.

File: addons/stock/models/res_company.py [hunk 5]
```diff
@@ -126,12 +141,20 @@ def create_missing_scrap_location(self):
...
+        company_todo_sequence = company_ids - company_has_scrap_seq
+        company_todo_sequence._create_scrap_sequence()
+
     def _create_per_company_locations(self):
         self.ensure_one()
         self._create_transit_location()
         self._create_inventory_loss_location()
         self._create_production_location()
         self._create_scrap_location()
+        self._create_scrap_sequence()
```
Comment (sle-odoo) [inline addons/stock/models/res_company.py]: not the right place

Comment (sle-odoo) [reply addons/stock/models/res_company.py]: maybe put it directly in _create_scrap_location, this way it is called when installing stock (the method is called in data.xml)

cc @amoyaux if he has a better idea

File: addons/stock/models/res_company.py [hunk 6]
```diff
@@ -126,16 +141,27 @@ def create_missing_scrap_location(self):
...
+        company_todo_sequence._create_scrap_sequence()
+
     def _create_per_company_locations(self):
         self.ensure_one()
         self._create_transit_location()
         self._create_inventory_loss_location()
         self._create_production_location()
         self._create_scrap_location()
 
+    def _create_per_company_scrap_sequence(self):
```
Comment (sle-odoo) [inline addons/stock/models/res_company.py]: rename to _create_per_company_sequences to be similar to _create_per_company_locations

Comment (sle-odoo) [reply addons/stock/models/res_company.py]: ensure_one to be similar to _create_per_company_locations

File: addons/stock/views/stock_scrap_views.xml [hunk 7]
```diff
@@ -61,6 +61,7 @@
                                 <field name="origin"/>
                                 <field name="date_expected"/>
                                 <field name="picking_id" attrs="{'invisible': [('picking_id', '=', False)]}"/>
+                                <field name="company_id" groups="base.group_multi_company"/>
```
Comment (Whenrow) [inline addons/stock/views/stock_scrap_views.xml]: As you add the company_id on the scrap. We should set a domain on location_id and scrap_location_id to restrict on the scrap's company. 

The default fonction in the model (_get_default_scrap_location_id and _get_default_location_id) should take location of the company as well (or location without company).

File: addons/stock/models/stock_scrap.py [hunk 8]
```diff
@@ -44,11 +48,11 @@ def _get_default_location_id(self):
     move_id = fields.Many2one('stock.move', 'Scrap Move', readonly=True)
     picking_id = fields.Many2one('stock.picking', 'Picking', states={'done': [('readonly', True)]})
     location_id = fields.Many2one(
-        'stock.location', 'Location', domain="[('usage', '=', 'internal')]",
+        'stock.location', 'Location', domain="[('usage', '=', 'internal'),('company_id', '=', company_id)]",
```
Comment (sle-odoo) [inline addons/stock/models/stock_scrap.py]: does it work?

File: addons/stock/models/stock_scrap.py [hunk 9]
```diff
@@ -12,7 +12,10 @@ class StockScrap(models.Model):
     _description = 'Scrap'
 
     def _get_default_scrap_location_id(self):
-        return self.env['stock.location'].search([('scrap_location', '=', True), ('company_id', 'in', [self.env.company.id, False])], limit=1).id
+        return self.env['stock.location'].search([
+            ('scrap_location', '=', True),
+            ('company_id', '=', self.env.company.id)
+        ], limit=1).id
```
Comment (sle-odoo) [inline addons/stock/models/stock_scrap.py]: why this change?

File: addons/mrp/models/mrp_workorder.py [hunk 10]
```diff
@@ -554,7 +554,10 @@ def button_scrap(self):
             'res_model': 'stock.scrap',
             'view_id': self.env.ref('stock.stock_scrap_form_view2').id,
             'type': 'ir.actions.act_window',
-            'context': {'default_workorder_id': self.id, 'default_production_id': self.production_id.id, 'product_ids': (self.production_id.move_raw_ids.filtered(lambda x: x.state not in ('done', 'cancel')) | self.production_id.move_finished_ids.filtered(lambda x: x.state == 'done')).mapped('product_id').ids},
+            'context': {'default_company_id': self.production_id.company_id.id,
+                        'default_workorder_id': self.id,
+                        'default_production_id': self.production_id.id,
+                        'product_ids': (self.production_id.move_raw_ids.filtered(lambda x: x.state not in ('done', 'cancel')) | self.production_id.move_finished_ids.filtered(lambda x: x.state == 'done')).mapped('product_id').ids},
             # 'context': {'product_ids': self.move_raw_ids.filtered(lambda x: x.state not in ('done', 'cancel')).mapped('product_id').ids + [self.production_id.product_id.id]},
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_workorder.py]: remove the comment please

File: addons/mrp/views/stock_scrap_views.xml [hunk 11]
```diff
@@ -6,8 +6,8 @@
         <field name="inherit_id" ref="stock.stock_scrap_form_view2"/>
         <field name="arch" type="xml">
             <field name="owner_id" position="after">
-                <field name="workorder_id" attrs="{'invisible': [('workorder_id', '=', False)]}"/>
-                <field name="production_id" attrs="{'invisible': [('production_id', '=', False)]}"/>
+                <field name="workorder_id" domain="[('production_id.company_id', '=', company_id)]" attrs="{'invisible': [('workorder_id', '=', False)]}"/>
```
Comment (sle-odoo) [inline addons/mrp/views/stock_scrap_views.xml]: wouldn't domain="[('production_id', '=', production_id)]" work?

File: addons/mrp/views/stock_scrap_views.xml [hunk 12]
```diff
@@ -17,8 +17,8 @@
         <field name="inherit_id" ref="stock.stock_scrap_form_view"/>
         <field name="arch" type="xml">
             <field name="owner_id" position="after">
-                <field name="workorder_id" attrs="{'invisible': [('workorder_id', '=', False)]}"/>
-                <field name="production_id" attrs="{'invisible': [('production_id', '=', False)]}"/>
+                <field name="workorder_id" domain="[('production_id.company_id', '=', company_id)]" attrs="{'invisible': [('workorder_id', '=', False)]}"/>
```
Comment (sle-odoo) [inline addons/mrp/views/stock_scrap_views.xml]: production_id

File: addons/stock/models/stock_scrap.py (hunk 9 above)
Comment (sle-odoo) [reply addons/stock/models/stock_scrap.py]: i would remove this diff

File: addons/stock/models/stock_scrap.py [hunk 13]
```diff
@@ -36,19 +40,19 @@ def _get_default_location_id(self):
     tracking = fields.Selection('Product Tracking', readonly=True, related="product_id.tracking")
     lot_id = fields.Many2one(
         'stock.production.lot', 'Lot',
-        states={'done': [('readonly', True)]}, domain="[('product_id', '=', product_id)]")
+        states={'done': [('readonly', True)]}, domain="[('product_id', '=', product_id), ('company_id', '=', company_id)]")
```
Comment (sle-odoo) [inline addons/stock/models/stock_scrap.py]: no company_id field on the lot at the moment

File: addons/stock/models/stock_scrap.py [hunk 14]
```diff
@@ -36,19 +40,19 @@ def _get_default_location_id(self):
...
     package_id = fields.Many2one(
         'stock.quant.package', 'Package',
         states={'done': [('readonly', True)]})
```
Comment (sle-odoo) [inline addons/stock/models/stock_scrap.py]: we do have a company_id field on stock quant package though

File: addons/stock/views/stock_scrap_views.xml [hunk 15]
```diff
@@ -53,15 +53,16 @@
                                 </div>
                             </group>
                             <group>
-                                <field name="lot_id" attrs="{'invisible': [('tracking', '=', 'none')], 'required': [('tracking', '!=', 'none')]}" groups="stock.group_production_lot"/>
+                                <field name="lot_id" attrs="{'invisible': [('tracking', 'in', ['none', False])], 'required': [('tracking', '!=', 'none')]}" groups="stock.group_production_lot"/>
```
Comment (sle-odoo) [inline addons/stock/views/stock_scrap_views.xml]: i don't get this change

File: addons/stock/views/stock_scrap_views.xml [hunk 16]
```diff
@@ -53,15 +53,16 @@
...
                                 <field name="tracking" invisible="1"/>
                                 <field name="package_id" groups="stock.group_tracking_lot"/>
                                 <field name="owner_id" groups="stock.group_tracking_owner"/>
                                 <field name="location_id" options="{'no_create': True}" groups="stock.group_stock_multi_locations"/>
                                 <field name="scrap_location_id" options="{'no_create': True}" groups="stock.group_stock_multi_locations"/>
                                 <field name="origin"/>
-                                <field name="date_expected"/>
+                                <field name="date_expected" attrs="{'readonly': [('state', '=', 'done')]}"/>
                                 <field name="picking_id" attrs="{'invisible': [('picking_id', '=', False)]}"/>
+                                <field name="company_id" groups="base.group_multi_company" attrs="{'readonly': [('state', '=', 'done')]}"/>
```
Comment (sle-odoo) [inline addons/stock/views/stock_scrap_views.xml]: the attrs is useless i think

**Comment by sle-odoo** [comment]
robodoo r+
